### PR TITLE
ur_modern_driver refactor

### PR DESCRIPTION
**Pull request by Zagitta**
Hello Thomas, here's the final code for the complete refactor as we've previously discussed.
During the 11/7 ROS Industrial meeting I asked about their preferred way to handle the code adoption and they greatly preferred if the changes could be merge into this repository to avoid having to redirect people elsewhere and cause more confusion than the old `ur_driver` already provides.

If you don't have time to handle the merging and issues that'll most likely crop up when people start using the new code you can add me as maintainer to the repository and I will do my best to handle it over the next month or two until the adoption ROS Industrial procedure has completed.

To get everyone else up to speed here's a rundown of everything that has happened in this rewrite:
- Implemented support for Kinetic (Indigo naturally still works, Jade is untested) #58 #82 
- Adopted clang-format with ROS specific style file
- Fixed issues #98 and #88 
- Implemented RT digital IO #50 
- Implemented joint temperature #81 
- Implemented parsing of tool analog values #83, still needs publishing.
- Implemented an enable robot service at `/ur_driver/robot_enable` that must be called on startup before the driver will accept commands. It must also be called if the robot ever enters a protective or emergency stop mode to ensure controllers acknowledge the stop mode.
- Added unit testing for message parsing (more work to be done here)

Besides the general list of changes above the driver now has a more sensible software design in the form of 2 producer/consumer work queues called pipelines, one for real time messages and one for the rest.
Each pipeline consists of a producer that received data from the robot, parses it and dumps the parsed messages into the queue which a consumer then picks up and runs through the various sub-consumers that publishes the messages.

The code has so far only been tested on a live UR5 v1.8, but has been developed against URSim v3.3.4 and should therefore work but it remains to be seen.

Last but not least I believe most of the blockers for #18 to happen have been resolved, the primary thing lacking now is better documentation but I'll try to get that done after my thesis defense.

Best regards, Simon R.

### CODE REVIEW

**Comment by v4hn** [comment]
Juchu! 😄 
Apparently there are conflicts with current upstream that will have to be resolved before this can be merged.

> Implemented an enable robot service at /ur_driver/robot_enable that must be called on startup before the driver will accept commands. It must also be called if the robot ever enters a protective or emergency stop mode to ensure controllers acknowledge the stop mode.

Is this optional? Otherwise this is a pretty big change in workflow that would have to be accounted for in all current setups.

**Comment by jonbinney** [comment]
Very impressive - @Zagitta if you can fix the merge conflicts I'll try this out on one of our UR10s and give some feedback.

**Comment by jonbinney** [comment]
Compiles for me, although I get a bunch of warnings about string formatting:
```
In file included from /home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/factory.h:9:0,
                 from /home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:17:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h: In instantiation of ‘bool URStateParser<RMD, MBD>::parse(BinParser&, std::vector<std::unique_ptr<StatePacket, st
d::default_delete<StatePacket> >, std::allocator<std::unique_ptr<StatePacket, std::default_delete<StatePacket> > > >&) [with RMD = RobotModeData_V3_2; MBD = MasterBoardData_V3_2]’:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:158:1:   required from here
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h:78:892: warning: format ‘%d’ expects argument of type ‘int’, but argument 8 has type ‘package_type’ [-Wformat=]
In file included from /home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/factory.h:9:0,
                 from /home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:17:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h:86:892: warning: format ‘%d’ expects argument of type ‘int’, but argument 8 has type ‘package_type’ [-Wformat=]
In file included from /home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/factory.h:9:0,
                 from /home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:17:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h: In instantiation of ‘bool URStateParser<RMD, MBD>::parse(BinParser&, std::vector<std::unique_ptr<StatePacket, st
d::default_delete<StatePacket> >, std::allocator<std::unique_ptr<StatePacket, std::default_delete<StatePacket> > > >&) [with RMD = RobotModeData_V3_0__1; MBD = MasterBoardData_V3_0__1]’:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:158:1:   required from here
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h:78:892: warning: format ‘%d’ expects argument of type ‘int’, but argument 8 has type ‘package_type’ [-Wformat=]
In file included from /home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/factory.h:9:0,
                 from /home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:17:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h:86:892: warning: format ‘%d’ expects argument of type ‘int’, but argument 8 has type ‘package_type’ [-Wformat=]
In file included from /home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/factory.h:9:0,
                 from /home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:17:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h: In instantiation of ‘bool URStateParser<RMD, MBD>::parse(BinParser&, std::vector<std::unique_ptr<StatePacket, st
d::default_delete<StatePacket> >, std::allocator<std::unique_ptr<StatePacket, std::default_delete<StatePacket> > > >&) [with RMD = RobotModeData_V1_X; MBD = MasterBoardData_V1_X]’:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:158:1:   required from here
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h:78:892: warning: format ‘%d’ expects argument of type ‘int’, but argument 8 has type ‘package_type’ [-Wformat=]
In file included from /home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/factory.h:9:0,
                 from /home/jbinney/ws/ur_modern/src/ur_modern_driver/src/ros_main.cpp:17:
/home/jbinney/ws/ur_modern/src/ur_modern_driver/include/ur_modern_driver/ur/state_parser.h:86:892: warning: format ‘%d’ expects argument of type ‘int’, but argument 8 has type ‘package_type’ [-Wformat=]
```

**Comment by ThomasTimm** [comment]
Good job Simon.

As soon as the few bugs have been fixed, I would propose to merge it into ROS-I where it belongs, instead of this repo which was only ever intended to be a temporary thing until such a merge was completed.

**Comment by v4hn** [comment]
Hey @Zagitta , @ThomasTimm .
We tested the proposed pull-request today in our lab and encountered a number of minor runtime issues.
Mostly teething problems. 😃
Please see [henhenhen's pull-request to branch](https://github.com/Zagitta/ur_modern_driver/pull/1) for details that should definitely be included.

The last thing I'm concerned about (we forgot about it today) is the `robot_enable` service.
We would like to be able to disable this from our launch files.
I fully see the need for this *as a feature*, but in R&D this can be quite inconvenient.

**Comment by Zagitta** [comment]
Thank you Thomas, I successfully completed the defense last Tuesday which is why I have time to get these things looked at now.

I tend to agree that this should only be a temporary repository, however from what I understood of the last ROS Industrial meeting they'd like use Github features to transfer your repository to their Github organization such that redirection links and such will be setup automatically to avoid further confusion between all the different ur_driver versions. 
From that perspective I think it makes sense to merge these changes, however because I know you're busy I think for now pull requests and issues regarding this refactor/rewrite should take place over in my fork (like @henhenhen has been kind enough to do) and then when we're ready can merge here and transfer to ROS Industrial organization.

Yes my apologies for not getting back to you sooner @v4hn, I'm currently working on making the `robot_enable` service have 3 configurable states that should satisfy everyone:
- Completely disabled, no calls required.
- Startup only, an initial call is required to "arm" the driver.
- Startup and faults, as above but must also be "re-armed" if the robot enters any protective or safety stop modes.

**Comment by v4hn** [comment]
> Yes my apologies for not getting back to you sooner @v4hn, I'm currently working on making the robot_enable service have 3 configurable states that should satisfy everyone:

👍 

Please have another look at my comment above, I added more content afterwards (tcp connections)

**Comment by Zagitta** [comment]
@v4hn During development I got the same error but suspect it had something todo with my Docker setup however it turns out `SO_REUSEADDR` was just getting set after binding rather than before, I have pushed a fix for this.
Regarding the WAITING and FIN states I'm not sure, do you know if this happens with the old version as well? The ActionServer basically opens and closes a connection per goal it receives (same for both old and new version) so it might be an inherent thing when sending it new goals quite rapidly?

**Comment by jonbinney** [comment]
Just tried this out on our UR10 on kinetic/16.04 and it worked great!

**Comment by BrettHemes** [comment]
Tried this out with a UR10 on 3.1.18024 and it seems to be working.  Some things to note:
- As @v4hn mentioned, the required call to the `robot_enable` is a bit inconvenient (as the robot is generally already enabled via pendant) but nothing we can't work around.
- The `URScript` topic seems to be missing now?.  Is this indeed gone?  If so, the README and pull request notes should be updated.

**Comment by v4hn** [comment]
On Tue, Aug 08, 2017 at 02:31:27PM +0000, Brett Hemes wrote:
> - As @v4hn mentioned, the required call to the `robot_enable` is a bit inconvenient

We are still waiting for the promised patch. :)

**Comment by v4hn** [comment]
We implemented the requested patch ourselves and it got merged into this request two weeks ago.
(https://github.com/Zagitta/ur_modern_driver/pull/2)

**Comment by ahundt** [comment]
@Zagitta can you add a tutorial to your driver? 

It looks great but I'm not sure how to set it up. It looks like there are some steps in either c++ / python you need to take for it to work and for it to change modes, but I'm not sure exactly what I need to write and/or activate, or what the advantage of ros_control is, for example.

**Comment by Zagitta** [comment]
@ahundt please don't hijack this PR with support requests.
Furthermore my refactor is a drop in replacement for this one and the setup is the same as described in Readme.md and an example of how to use the driver with python can be found in https://github.com/ThomasTimm/ur_modern_driver/blob/master/test_move.py
If these instructions don't make sense I suggest you go through the tutorials at http://wiki.ros.org/ROS/Tutorials

**Comment by ahundt** [comment]
@Zagitta Sorry, I first tried to create an issue on your repository but you have them disabled there. Since my change request would ultimately change this pull request this is also a reasonable location to post. But really a link to the "hello world" of ros tutorials is unhelpful because my questions were very specific, and you have these "[activation modes](https://github.com/Zagitta/ur_modern_driver/commit/231840fabff8ed5671f0eaabf49623c4ea516206)" which are neither here nor in either of the tutorials you linked.

**Comment by Zagitta** [comment]
@ahundt I read your initial comment while on public transport and assumed you meant the different controller modes and basics of how to launch the driver and communicate with it in python, I apologize for the rudeness.
Documenting the activation modes is indeed a very reasonable request. The default behavior is to not require any activation and you can simply ignore it while the two other options means you have to actively start the driver with manual input by calling the `ur_driver/robot_enable` service for example from the terminal with the command `rosservice call /ur_driver/robot_enable`
It might seem rather silly but it was done to improve safety in cases where the driver crashes and is restarted but some someone keeps publishing goals in the meanwhile which might be vastly different to the robot position and make it swing wildly.

I hope that clears it up, I'll try to add it to the Readme.

File: src/ros_main.cpp [hunk 1]
```diff
@@ -65,6 +67,7 @@ bool parse_args(ProgArgs &args)
   ros::param::param(PREFIX_ARG, args.prefix, std::string());
   ros::param::param(BASE_FRAME_ARG, args.base_frame, args.prefix + "base_link");
   ros::param::param(TOOL_FRAME_ARG, args.tool_frame, args.prefix + "tool0_controller");
+  ros::param::param(TCP_LINK_ARG, args.tcp_link, args.prefix + "ee_link");
```
Comment (gavanderhoorn) [inline src/ros_main.cpp]: Could we use `tool0` for this? `ee_link`'s pose does not correspond with the UR controller's local coordinate frame in the `ur_description` xacros.

Comment (v4hn) [reply src/ros_main.cpp]: Argh, @gavanderhoorn is right. I mixed up the two frames in the commit.
`args.prefix+ "tool0"` is the default I was going for.
Could you please change it @Zagitta ?

**Comment by mrjogo** [comment]
@Zagitta I've run into two issues using your fork that I didn't see upstream. Where should I report/discuss them? (specifically: immediate consumer overflow upon starting with `use_ros_control=True`, and a trajectory that should trigger a UR3 protective stop instead results in the UR3 not stopping until the estop is hit) Thanks!

**Comment by Zagitta** [comment]
@mrjogo I have enabled Issues on my repository at https://github.com/Zagitta/ur_modern_driver/issues now, feel free to report the error there. I was actually not aware I had managed to disabled issues.

**Comment by ahundt** [comment]
@Zagitta Any plans to merge the Safe Trajectory Follower implementation https://github.com/Zagitta/ur_modern_driver/pull/9 into this PR and update to master?

It is kind of tricky to jump around and figure out which PR to use at this point. :-)

**Comment by gavanderhoorn** [comment]
All of the current PRs are being evaluated and merged into their respective repositories when deemed appropriate.

The situation is a bit complex right now but that will be resolved.

As to Zagitta/ur_modern_driver#9: some issues / unexpected behaviour has been identified which will need to be resolved before we can think about merging that in.

Unfortunately that means that if you want to use it right now you'll have to build from the branch itself.

**Comment by dniewinski** [comment]
Just an FYI: Just tested this on a UR5 with 3.5.4.10845 (Apr 13 2018) and Kinetic with MoveIt! and didn't notice any issues.
screenshot from 2018-07-12 11-56-43

**Comment by shaun-edwards** [comment]
We are actively using the refactored branch and have exercised the following functionality without any issue:

- Joint states
- IO
- Action based motion

+1 for merging these changes.

**Comment by gavanderhoorn** [comment]
As an update: as part of the repository maintenance activities that will be initiated next month (September), this PR will be merged into this repository.

If anyone has any serious objections to this, now would be the time to make them heard.

We will not request additional changes from the submitter (@Zagitta), so enhancements or small fixups will be done after the initial merge.

**Comment by gavanderhoorn** [comment]
Finally got this merged. :fireworks: 

Thanks @Zagitta for all the work and thanks to the many reviewers and users of this PR for their feedback.

:+1: 

As I've written in earlier comments and on other PRs: there have been various reasons for why some PRs have been open for a rather long time and I do apologise for that. Even though it may seem that way, this is not indicative of disinterest in the feature or fix.

**Comment by v4hn** [comment]
Thank you so much for your time and commitment @gavanderhoorn !
Happy ROSCon :1st_place_medal:

**Comment by ahundt** [comment]
This is awesome, thanks! 👍

**Comment by AndyZe** [comment]
I'm just noticing, the /joint_speed topic is gone from this refactored version. I very much liked that topic for streaming poses to the robot in real time. Would it be difficult to re-enable /joint_speed?

I know /ur_script can also handle streaming in real time, but it's not compatible with other robots.

**Comment by miguelprada** [comment]
I never used the `/joint_speed` topic myself, but I believe you should achieve similar functionality by properly configuring a `velocity_controllers/JointGroupVelocityController`. This should be even more compatible with other robots than the `/joint_speed` topic.

**Comment by AndyZe** [comment]
@miguelprada A ros_control question for you. (Maybe we can move the discussion to email or answers.ros.org? I can't do google groups).

Is there a fundamental reason why I can't stream velocities to the default ros_control type, velocity_controllers/JointTrajectoryController? I attempted this earlier today, and got the following error:

> [ WARN] [1539981427.410135432]: Dropping first 1 trajectory point(s) out of 30, as they occur before the current time.
First valid point will be reached in 0.002s.
[ERROR] [1539981427.411941084]: Unexpected error: No trajectory defined at current time. Please contact the package maintainer.


I can understand why some points were skipped, and I can overcome that issue easily. But I don't understand the error related to "no trajectory defined at current time."

I would like to avoid switching back and forth between different ros_control types because I need to plan trajectories as well as stream commands.

**Comment by gavanderhoorn** [comment]
@AndyZe: I'm not @miguelprada, but: the `JointTrajectory` action interface accepts `JointTrajectory` goals. Those are inherently position based, but do support velocity and acceleration constraints.

That interface is not meant for streaming velocities, and that is why you get that error (the action server sees a trajectory coming in with no values for the `position` field).

**Comment by gavanderhoorn** [comment]
> I would like to avoid switching back and forth between different ros_control types because I need to plan trajectories as well as stream commands. (And setting up a new control type has been hard.)

I don't see a way around this: `ros_control` supports controller switching and that would seem to allow you to do precisely what you want: when you need to stream velocities, switch to the controller @miguelprada suggested. For trajectory execution, switch to the `joint_trajectory_controller`.

**Comment by miguelprada** [comment]
@gavanderhoorn beat me to it :wink: 

I believe you are probably trying to send joint trajectories following the pattern that the previous `joint_speed` topic was expecting, i.e. a single `JointTrajectoryPoint` with only velocity/acceleration setpoints. This will never work with the `joint_trajectory_controller` which, as stated, is inherently position based.

That's irrespective of which underlying joint interface is used by the controller to track your trajectory, e.g. in `velocity_controllers/JointTrajectoryController` joint velocity commands are used internally by the controller to track your _position_ trajectory.

You can use `joint_trajectory_controller` to stream positions (single point trajectories), and that's precisely what `rqt_joint_trajectory_controller` does, but even this is kind of abusing the controller somehow.

**Comment by miguelprada** [comment]
@gavanderhoorn beat me to it :wink: 

I believe you are probably trying to send joint trajectories following the pattern that the previous `joint_speed` topic was expecting, i.e. a single `JointTrajectoryPoint` with only velocity/acceleration setpoints. This will never work with the `joint_trajectory_controller` which, as stated, is inherently position based.

That's irrespective of which underlying joint interface is used by the controller to track your trajectory, e.g. in `velocity_controllers/JointTrajectoryController` joint velocity commands are used internally by the controller to track your _position_ trajectory.

You can use `joint_trajectory_controller` to stream positions (single point trajectories), and that's precisely what `rqt_joint_trajectory_controller` does, but even this is kind of abusing the controller somehow.

**Comment by miguelprada** [comment]
@gavanderhoorn beat me to it :wink: 

I believe you are probably trying to send joint trajectories following the pattern that the previous `joint_speed` topic was expecting, i.e. a single `JointTrajectoryPoint` with only velocity/acceleration setpoints. This will never work with the `joint_trajectory_controller` which, as stated, is inherently position based.

That's irrespective of which underlying joint interface is used by the controller to track your trajectory, e.g. in `velocity_controllers/JointTrajectoryController` joint velocity commands are used internally by the controller to track your _position_ trajectory.

You can use `joint_trajectory_controller` to stream positions (single point trajectories), and that's precisely what `rqt_joint_trajectory_controller` does, but even this is kind of abusing the controller somehow.

**Comment by miguelprada** [comment]
@gavanderhoorn beat me to it :wink: 

I believe you are probably trying to send joint trajectories following the pattern that the previous `joint_speed` topic was expecting, i.e. a single `JointTrajectoryPoint` with only velocity/acceleration setpoints. This will never work with the `joint_trajectory_controller` which, as stated, is inherently position based.

That's irrespective of which underlying joint interface is used by the controller to track your trajectory, e.g. in `velocity_controllers/JointTrajectoryController` joint velocity commands are used internally by the controller to track your _position_ trajectory.

You can use `joint_trajectory_controller` to stream positions (single point trajectories), and that's precisely what `rqt_joint_trajectory_controller` does, but even this is kind of abusing the controller somehow.

**Comment by miguelprada** [comment]
@gavanderhoorn beat me to it :wink: 

I believe you are probably trying to send joint trajectories following the pattern that the previous `joint_speed` topic was expecting, i.e. a single `JointTrajectoryPoint` with only velocity/acceleration setpoints. This will never work with the `joint_trajectory_controller` which, as stated, is inherently position based.

That's irrespective of which underlying joint interface is used by the controller to track your trajectory, e.g. in `velocity_controllers/JointTrajectoryController` joint velocity commands are used internally by the controller to track your _position_ trajectory.

You can use `joint_trajectory_controller` to stream positions (single point trajectories), and that's precisely what `rqt_joint_trajectory_controller` does, but even this is kind of abusing the controller somehow.

**Comment by miguelprada** [comment]
@gavanderhoorn beat me to it :wink: 

I believe you are probably trying to send joint trajectories following the pattern that the previous `joint_speed` topic was expecting, i.e. a single `JointTrajectoryPoint` with only velocity/acceleration setpoints. This will never work with the `joint_trajectory_controller` which, as stated, is inherently position based.

That's irrespective of which underlying joint interface is used by the controller to track your trajectory, e.g. in `velocity_controllers/JointTrajectoryController` joint velocity commands are used internally by the controller to track your _position_ trajectory.

You can use `joint_trajectory_controller` to stream positions (single point trajectories), and that's precisely what `rqt_joint_trajectory_controller` does, but even this is kind of abusing the controller somehow.

**Comment by miguelprada** [comment]
@gavanderhoorn beat me to it :wink: 

I believe you are probably trying to send joint trajectories following the pattern that the previous `joint_speed` topic was expecting, i.e. a single `JointTrajectoryPoint` with only velocity/acceleration setpoints. This will never work with the `joint_trajectory_controller` which, as stated, is inherently position based.

That's irrespective of which underlying joint interface is used by the controller to track your trajectory, e.g. in `velocity_controllers/JointTrajectoryController` joint velocity commands are used internally by the controller to track your _position_ trajectory.

You can use `joint_trajectory_controller` to stream positions (single point trajectories), and that's precisely what `rqt_joint_trajectory_controller` does, but even this is kind of abusing the controller somehow.

**Comment by AndyZe** [comment]
Alright, thanks guys. Actually my `JointTrajectoryPoint` did include a position but I guess a `velocity_controllers/JointTrajectoryController` isn't suitable for single-point streaming, anyway (as miguelprada alluded to). More in this [github issue](https://github.com/ros-controls/ros_controllers/issues/143).

**Comment by miguelprada** [comment]
> Alright, thanks guys. Actually my JointTrajectoryPoint did include a position but I guess a velocity_controllers/JointTrajectoryController isn't suitable for single-point streaming, anyway (as miguelprada alluded to). More in this github issue.

I never tried to imply that the `velocity_controllers/JointTrajectoryController` is ill-suited for single-point streaming. I often use `velocity_controllers/JointTrajectoryController` together with `rqt_joint_trajectory_controller` to jog the robot around, and if I'm not wrong this is effectively streaming single trajectory points.

How are you exactly filling your trajectory message? In particular, how are you timestamping the message and each of the points (i.e. the `time_from_start` field)? I would suggest you play around with `rqt_joint_trajectory_controller` and try to figure out why this works and your code doesn't.

**Comment by gavanderhoorn** [comment]
@AndyZe @miguelprada: could we create a new issue for discussing the velocity streaming approaches with this driver? I welcome the discussion, but let's not have it in the comment section of an already merged PR.

**Pull request merged**